### ISSUE TITLE
Image.SetKill(true) does not work on DzSave

### ISSUE THREAD

**Issue by ax-meyer** [kleisauke/net-vips#179]
I am having trouble aborting a DzSave call.

Sample Code:

```
using NetVips;

Console.WriteLine("Starting");

// setup parameters
var testImage = @"sample_640×426.tif";
var savePath = "output";
var saveName = savePath + "/" + "testPyramid";

if (!Directory.Exists(savePath))
    Directory.CreateDirectory(savePath);

// open image
using var image = Image.NewFromFile(testImage);

// start creation of pyramid in background task
VipsException? vipsException = null;
var createPyramidTask = Task.Run(() =>
{
    {
        try
        {
            image.Dzsave(filename: saveName,
                tileSize: 10,
                overlap: 0,
                suffix: ".tif",
                depth: Enums.ForeignDzDepth.Onetile
            );
        }
        catch (VipsException ex)
        {
            vipsException = ex;
        }
    }
});

var start = DateTime.Now;

// wait one second and set kill flag. Wait until task returns.
await Task.Run(async () =>
{

    while (DateTime.Now - start < TimeSpan.FromSeconds(2))
        Thread.Sleep(10);

    image.SetKill(true);
    Console.WriteLine("killflag set");
    
    await createPyramidTask.ConfigureAwait(false);
});

Console.WriteLine("duration {0}", DateTime.Now - start);
Console.WriteLine("is killed? {0}", image.IsKilled());
Console.WriteLine("exception: {0}", vipsException?.Message);

Console.ReadKey();

Directory.Delete(savePath, true);
```

Sample image is attached: [sample_640×426.zip](https://github.com/kleisauke/net-vips/files/9584307/sample_640x426.zip)
This gets the following output in the console:
```
Starting
killflag set
duration 00:00:11.4808162
is killed? True
exception:

```
So it says it was killed, but instead of terminating after abount 1 second, it more than 10 times as long. And looking in the output folder, the whole image pyramid was actually generated.

Does `SetKill()` not work on `DzSave`?

**Comment by kleisauke** [comment]
I was unable to reproduce this with the [`ImagePyramid` sample](https://github.com/kleisauke/net-vips/blob/02be3d67c4fe7f83069d3ec4f10210db8b7f216b/samples/NetVips.Samples/Samples/ImagePyramid.cs) that passed the `CancellationToken` to `SetProgress`:
https://github.com/kleisauke/net-vips/blob/02be3d67c4fe7f83069d3ec4f10210db8b7f216b/samples/NetVips.Samples/Samples/ImagePyramid.cs#L24-L25

Simplifying the example program above also works without any issues:
```csharp
// Build test image
using var im = Image.NewFromFile("images/sample2.v", access: Enums.Access.Sequential);
using var test = im.Replicate(50, 50);

var cts = new CancellationTokenSource();
cts.CancelAfter(2000);

await Task.Run(async () =>
{
    while (!cts.IsCancellationRequested)
    {
        Console.Write("*");

        await Task.Delay(100);
    }

    test.SetKill(true);
    Console.WriteLine("\nImage evaluation blocked");
}, cts.Token);

try
{
    // Save image pyramid
    test.Dzsave("images/image-pyramid",
        tileSize: 10,
        overlap: 0,
        suffix: ".tif",
        depth: Enums.ForeignDzDepth.Onetile
    );
}
catch (VipsException ex)
{
    // Catch and log the VipsException,
    // because we may block the evaluation of this image
    Console.WriteLine("\n" + ex.Message);
}
```

**Comment by kleisauke** [comment]
I was able to reproduce this. On my PC, it takes 18 seconds to see the effects of `vips_image_set_kill()` in the above example.

Increasing the `tileSize` to 254 or applying this patch:
```diff
--- a/libvips/foreign/dzsave.c
+++ b/libvips/foreign/dzsave.c
@@ -849,11 +849,15 @@ write_image( VipsForeignSaveDz *dz,
 	GsfOutput *out, VipsImage *image, const char *format )
 {
 	VipsObjectClass *class = VIPS_OBJECT_GET_CLASS( dz );
+	VipsForeignSave *save = VIPS_FOREIGN_SAVE( dz );
 
 	VipsImage *t;
 	void *buf;
 	size_t len;
 
+	if( vips_image_iskilled( save->in ) )
+		return( -1 );
+
 	/* We need to block progress signalling on individual image write, so
 	 * we need a copy of the tile in case it's shared (eg. associated
 	 * images).
```

seems to fix this.

/cc @jcupitt

**Comment by jcupitt** [comment]
Oh, good idea. The kill flag is checked in the thing that loops down the image, but of course that's only checked once per line of tiles.

It could maybe go into strip_work() -- that might be even quicker. I'll add this to 8.13. Thanks!

**Comment by ax-meyer** [comment]
> I was able to reproduce this. On my PC, it takes 18 seconds to see the effects of `vips_image_set_kill()` in the above example.
> 
> Increasing the `tileSize` to 254 or applying this patch:

The tiny `tileSize` was just for demonstration purposes here to provoke a long running `Dzsave` on the small image I attached. I am actually using a `tileSize` of 2000 pixels, on images in the order of 80000 x 60000 pixels.

For the moment, I wrapped the creation step into its own executable and start in in a separate process from my main program. And this process can be hard killed if needed.

But I am looking forward to the patch - will definitely make my life easier.
